### ISSUE TITLE
Add Multiple ML Algorithms to Actions Module for Enhanced Decision Making

### ISSUE THREAD

**Issue by csmangum** [Dooders/AgentFarm#270]
### Current Action Decision-Making Analysis

After examining the actions module, the current decision-making approaches include:

**Existing Patterns:**
- **DQN-based**: `BaseDQNModule` with neural networks for Q-learning
- **Rule-based**: Hardcoded logic in various action modules
- **Heuristic-based**: Simple distance calculations, threshold checks
- **Probabilistic**: Action selection via `SelectModule` with weighted probabilities

**Current Limitations:**
- Limited algorithm diversity for decision-making
- No traditional ML approaches available
- Difficult to compare different decision-making strategies
- No ensemble methods or hybrid approaches

### Proposed Enhancement: Multi-Algorithm Action Decision Making

**Objective:** Implement multiple machine learning algorithms within the actions module to provide diverse decision-making capabilities and enable algorithm comparison.

#### 1. **Multi-Layer Perceptron (MLP) Classifier/Regressor**
```python
class MLPActionSelector:
    """MLP-based action selection using supervised learning approach."""
    - Input: Agent state features
    - Output: Action probabilities or Q-values
    - Architecture: Configurable hidden layers with dropout
    - Training: Supervised learning on expert demonstrations or self-play data
```

#### 2. **Support Vector Machine (SVM)**
```python
class SVMActionSelector:
    """SVM-based action classification for discrete action spaces."""
    - Kernel options: RBF, linear, polynomial
    - Multi-class classification for action selection
    - Probability outputs for exploration
```

#### 3. **Random Forest Classifier**
```python
class RandomForestActionSelector:
    """Ensemble learning approach using decision trees."""
    - Configurable number of trees
    - Feature importance analysis
    - Robust to overfitting
```

#### 4. **Gradient Boosting (XGBoost/LightGBM)**
```python
class GradientBoostActionSelector:
    """Gradient boosting for action selection."""
    - High performance on structured data
    - Feature importance and interpretability
    - Regularization options
```

#### 5. **Naive Bayes Classifier**
```python
class NaiveBayesActionSelector:
    """Probabilistic classifier for action selection."""
    - Fast training and prediction
    - Good baseline performance
    - Works well with categorical features
```

#### 6. **K-Nearest Neighbors**
```python
class KNNActionSelector:
    """Instance-based learning for action selection."""
    - Distance-weighted action selection
    - Memory-based approach
    - Good for non-stationary environments
```

#### 1. **Unified Algorithm Interface**
```python
class ActionAlgorithm(ABC):
    """Abstract base class for all action selection algorithms."""
    
    @abstractmethod
    def select_action(self, state: np.ndarray) -> int:
        """Select action given current state."""
        pass
    
    @abstractmethod
    def train(self, states: np.ndarray, actions: np.ndarray, rewards: np.ndarray):
        """Train the algorithm on experience data."""
        pass
    
    @abstractmethod
    def predict_proba(self, state: np.ndarray) -> np.ndarray:
        """Predict action probabilities for exploration."""
        pass
    
    @abstractmethod
    def save_model(self, path: str):
        """Save model to disk."""
        pass
    
    @abstractmethod
    def load_model(self, path: str):
        """Load model from disk."""
        pass
```

#### 2. **Algorithm Registry**
```python
class AlgorithmRegistry:
    """Registry for managing different action algorithms."""
    
    _algorithms = {
        'mlp': MLPActionSelector,
        'svm': SVMActionSelector,
        'random_forest': RandomForestActionSelector,
        'gradient_boost': GradientBoostActionSelector,
        'naive_bayes': NaiveBayesActionSelector,
        'knn': KNNActionSelector,
        'dqn': DQNActionSelector,  # Existing DQN wrapper
    }
```

#### 3. **Configuration System**
Extend existing config to support algorithm selection:

```python
class ActionAlgorithmConfig(BaseModel):
    algorithm_type: str = Field(default="mlp")
    algorithm_params: Dict[str, Any] = Field(default_factory=dict)
    feature_engineering: List[str] = Field(default_factory=list)
    ensemble_size: int = Field(default=1)
    use_exploration_bonus: bool = Field(default=True)
    
    @field_validator("algorithm_type")
    def validate_algorithm(cls, v):
        valid_types = ['mlp', 'svm', 'random_forest', 'gradient_boost', 
                      'naive_bayes', 'knn', 'dqn']
        if v not in valid_types:
            raise ValueError(f"Algorithm must be one of: {valid_types}")
        return v
```

#### 1. **Action Module Integration**
Update existing action modules to support algorithm selection:
- `MoveModule` - Use different algorithms for movement decisions
- `AttackModule` - ML-based attack strategy selection
- `GatherModule` - Resource gathering optimization
- `SelectModule` - Meta-action selection using ML

#### 2. **Feature Engineering**
```python
class FeatureEngineer:
    """Feature engineering for ML algorithms."""
    
    def extract_features(self, agent, environment) -> np.ndarray:
        """Extract relevant features for ML algorithms."""
        features = []
        
        # Agent state features
        features.extend([
            agent.health / 100.0,
            agent.resources / 100.0,
            agent.position.x / environment.width,
            agent.position.y / environment.height,
        ])
        
        # Environmental features
        features.extend(self._extract_environmental_features(agent, environment))
        
        # Social features
        features.extend(self._extract_social_features(agent, environment))
        
        return np.array(features)
```

### Dependencies to Add

```python
# Add to requirements.txt
scikit-learn>=1.3.0  # For traditional ML algorithms
xgboost>=1.7.0       # Gradient boosting
lightgbm>=3.3.0      # Alternative gradient boosting
pandas>=1.5.0        # Data manipulation for ML
joblib>=1.2.0        # Model serialization
```

#### 1. **Experience Collection**
```python
class ExperienceCollector:
    """Collect training data for ML algorithms."""
    
    def collect_episode(self, agent, environment):
        """Collect state-action-reward tuples from agent behavior."""
        episode_data = []
        
        for step in range(max_steps):
            state = self.feature_engineer.extract_features(agent, environment)
            action = agent.select_action()
            reward = environment.step(agent, action)
            
            episode_data.append((state, action, reward))
        
        return episode_data
```

#### 2. **Algorithm Training**
```python
class AlgorithmTrainer:
    """Train ML algorithms on collected experience."""
    
    def train_algorithm(self, algorithm: ActionAlgorithm, 
                       training_data: List[Tuple[np.ndarray, int, float]]):
        """Train specific algorithm on collected data."""
        states, actions, rewards = zip(*training_data)
        
        # Convert to numpy arrays
        X = np.array(states)
        y_actions = np.array(actions)
        y_rewards = np.array(rewards)
        
        # Train the algorithm
        algorithm.train(X, y_actions, y_rewards)
```

#### 1. **Benchmarking System**
```python
class AlgorithmBenchmark:
    """Compare performance of different algorithms."""
    
    def run_comparison(self, algorithms: List[str], 
                      environments: List[str], 
                      metrics: List[str]):
        """Run comprehensive algorithm comparison."""
        results = {}
        
        for algo_name in algorithms:
            for env_name in environments:
                performance = self._evaluate_algorithm(algo_name, env_name, metrics)
                results[f"{algo_name}_{env_name}"] = performance
        
        return results
```

### Expected Outcomes

1. **Algorithm Diversity:** Multiple approaches for different problem types
2. **Performance Comparison:** Built-in benchmarking capabilities
3. **Interpretability:** Feature importance from tree-based methods
4. **Flexibility:** Easy switching between algorithms
5. **Research Enablement:** Foundation for ML-based agent research

### Files to Create/Modify

**New Files:**
- `farm/actions/algorithms/` - Directory for algorithm implementations
- `farm/actions/algorithms/base.py` - Abstract base classes
- `farm/actions/algorithms/mlp.py` - MLP implementation
- `farm/actions/algorithms/svm.py` - SVM implementation
- `farm/actions/algorithms/ensemble.py` - Tree-based algorithms
- `farm/actions/feature_engineering.py` - Feature extraction utilities
- `farm/actions/training/` - Training and data collection
- `farm/actions/benchmark/` - Algorithm comparison tools

**Modified Files:**
- `farm/actions/config.py` - Add algorithm configuration
- `farm/actions/base_dqn.py` - Integrate with new system
- Update all action modules to support algorithm selection

### Testing Requirements

- Unit tests for each algorithm implementation
- Integration tests with existing action modules
- Performance benchmarks comparing algorithms
- Feature engineering validation tests
- Training pipeline tests

### Related Issues

- **RL Integration**: Complement existing DQN with traditional ML approaches
- **Benchmarking**: Extend current benchmarking to include ML algorithms
- **Feature Engineering**: Develop comprehensive feature sets for agents

### Implementation Priority

1. **High Priority:** MLP, Random Forest, Feature Engineering
2. **Medium Priority:** SVM, Gradient Boosting, Naive Bayes  
3. **Low Priority:** KNN, Advanced ensemble methods

**Comment by csmangum** [comment]
@cursor is this issue related to #269

**Comment by csmangum** [comment]
@cursor go ahead and implement this issue
